uportal-war/src/main/java/org/jasig/portal/groups/pags/testers/AdHocGroupTester.java

```diff
@@ -0,0 +1,208 @@
+
+    private final Cache currentTests;
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+    private final Set<IEntityGroup> includes = new HashSet<>();
+    private final Set<IEntityGroup> excludes = new HashSet<>();
+    private final String groupsHash;
+
+    public AdHocGroupTester(IPersonAttributesGroupTestDefinition definition) {
+        Validate.notNull(definition.getIncludes());
+        Validate.notNull(definition.getExcludes());
+        this.includes.addAll(collectGroups(definition.getIncludes(), true));
+        this.excludes.addAll(collectGroups(definition.getExcludes(), false));
+        this.groupsHash = calcGroupsHash(definition.getIncludes(), definition.getExcludes());
+        ApplicationContext context = ApplicationContextLocator.getApplicationContext();
+        CacheManager cacheManager = context.getBean("cacheManager", CacheManager.class);
+        this.currentTests = cacheManager.getCache("org.jasig.portal.groups.pags.testers.AdHocGroupTester");
+    }
+
+    /*
+     * At some point, a person is being tested for group membership. During that test, the thread hits an ad hoc group
+     * tester. When that tester calls isDeepMemberOf, a test for group membership is triggered. During this call stack,
+     * the second call the the ad hoc group tester returns false. Assuming the group hierarchy is not itself recursive
+     * for the group containing the ad hoc group test, the test returns a usable value.
+     *
+     * If there is no caching and the second person object only exists for the recursive call, then the implementation
+     * works.
+     *
+     * Also, if the person object is cached and used twice, then the group key with the ad hoc tester is not added to
+     * the containing group keys during the recursion but is added (or not) after the test call returns positive.
+     */
+    @Override
+    public boolean test(IPerson person) {
+        String personHash = person.getEntityIdentifier().getKey() + groupsHash + Thread.currentThread().getId();
+        logger.debug("Entering test() for {}", personHash);
+        if (currentTests.getQuiet(personHash) != null) {
+            logger.warn("Returning from test() for {} due to recusion for person = {}", personHash, person.toString());
+            return false; // stop recursing
+        }
+        Element cacheEl = new Element(personHash, personHash);
+        currentTests.put(cacheEl);
+
+        IGroupMember gmPerson = findPersonAsGroupMember(person);
+        boolean pass = true;
+
+        Iterator<IEntityGroup> groups = includes.iterator();
+        while (pass && groups.hasNext()) {
+            IEntityGroup group = groups.next();
+            logger.debug("Checking if {} is a member of {}", person.getUserName(), group.getName());
+            if (!gmPerson.isDeepMemberOf(group)) {
+                logger.debug("!! {} is not a member of {}", person.getUserName(), group.getDescription());
+                pass = false;
```

Comment:
The goal for a positive test result is for the person to be a member of every includes group and not a member of every excludes group. By starting with pass = true, the first time a membership check fails, the while loops are exited/skipped. The excludes while loop checks `pass`, so if an includes fails it is skipped.

I will add documentation.